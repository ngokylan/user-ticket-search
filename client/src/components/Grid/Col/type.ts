import { ReactNode } from 'react';

export type ColResponsiveOption = number | object;
export type ColAlign =
  | null
  | 'auto'
  | 'start'
  | 'end'
  | 'center'
  | 'baseline'
  | 'stretch';

export type ColProps = {
  /** Content of col */
  children?: ReactNode;
  /** Span atribute of col */
  span?: number;
  /** If is auto */
  isAuto?: boolean;
  /** order atribute of col */
  order?: number;
  /** offset atribute of col */
  offset?: number;
  /**
   * align atribute of col, use one from the following:
   * ColAlign =  null | 'auto' | 'start' | 'end' | 'center' | 'baseline' | 'stretch'
   */
  align?: ColAlign;
  /** shrink atribute of col */
  shrink?: number;
  /** grow atribute of col */
  grow?: number;
  /** Fill atribute of col */
  isFill?: boolean;
  /** xs responsive atribute of col */
  xs?: ColResponsiveOption;
  /** sm responsive atribute of col */
  sm?: ColResponsiveOption;
  /** md responsive atribute of col */
  md?: ColResponsiveOption;
  /** lg responsive atribute of col */
  lg?: ColResponsiveOption;
  /** xl responsive atribute of col */
  xl?: ColResponsiveOption;
  /** xxl responsive atribute of col */
  xxl?: ColResponsiveOption;
  /** xxl responsive atribute of col */
  xxxl?: ColResponsiveOption;
  /** key atribute of col */
  key?: number | string;
  /** Col class name attribute, (optional) */
  className?: string;
  /** WCAG role */
  role?: string;
};

export type ColResponsivePropertiesType = {
  property: string;
  prefix: string;
};
