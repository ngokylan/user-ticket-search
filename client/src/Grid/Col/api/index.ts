import { getClass } from '../../../_lib/helper';
import { ColResponsivePropertiesType, ColProps } from '../type';

/**
 * Responsive properties
 */
export const responsiveProperties: ColResponsivePropertiesType[] = [
  { property: 'span', prefix: 'col-' },
  { property: 'auto', prefix: 'col-' },
  { property: 'order', prefix: 'order-' },
  { property: 'offset', prefix: 'offset-' },
  { property: 'alignSelf', prefix: 'align-self-' },
  { property: 'shrink', prefix: 'flex-shrink-' },
  { property: 'grow', prefix: 'flex-grow-' },
  { property: 'fill', prefix: 'flex-' },
];

/**
 *
 * @param block
 * @param props
 */
export function mapPropsToClass(block: string, props: ColProps) {
  const {
    className,
    isAuto,
    span,
    order,
    offset,
    align,
    shrink,
    grow,
    isFill,
    xs,
    sm,
    md,
    lg,
    xl,
    xxl,
    xxxl,
  } = props;
  return getClass('col', {
    [`${className}`]: className,
    [`${block}-auto`]: isAuto,
    [`col-${span}`]: span,
    [`order-${order}`]: order,
    [`offset-${offset}`]: offset,
    [`align-self-${align}`]: align,
    [`flex-shrink-${shrink}`]: shrink !== undefined,
    [`flex-grow-${grow}`]: grow !== undefined,
    [`flex-fill`]: isFill,
    [`${mapResponsive(block, 'xs', xs)}`]: xs,
    [`${mapResponsive(block, 'sm', sm)}`]: sm,
    [`${mapResponsive(block, 'md', md)}`]: md,
    [`${mapResponsive(block, 'lg', lg)}`]: lg,
    [`${mapResponsive(block, 'xl', xl)}`]: xl,
    [`${mapResponsive(block, 'xxl', xxl)}`]: xxl,
    [`${mapResponsive(block, 'xxxl', xxxl)}`]: xxxl,
  });
}

/**
 *
 * @param breakpoint
 * @param options
 */
export function mapResponsive(block: string, breakpoint: string, options: any) {
  if (options === null || options === undefined) {
    return '';
  }

  if (typeof options === 'number') {
    return `${block}-${breakpoint}-${options}`;
  }

  return responsiveProperties
    .map(item => {
      if (options[item.property] !== undefined) {
        return `${item.prefix}${breakpoint}-${options[item.property]}`;
      }
    })
    .join(' ');
}
