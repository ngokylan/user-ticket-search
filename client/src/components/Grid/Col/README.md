
Basic
```js
import Container from '../Container';
import Row from '../Row';

<Container>
  <Row>
    <Col className="demo-col">A</Col>
    <Col className="demo-col">B</Col>
    <Col className="demo-col">C</Col>
  </Row>
</Container>
```

Span
```js
import Container from '../Container';
import Row from '../Row';

<Container>
  <Row>
    <Col className="demo-col" span={6}>A</Col>
    <Col className="demo-col" span={12}>B</Col>
    <Col className="demo-col" span={6}>C</Col>
  </Row>
  <br />
  <Row>
    <Col className="demo-col" span={4}>A</Col>
    <Col className="demo-col" span={10}>B</Col>
    <Col className="demo-col" span={10}>C</Col>
  </Row>
  <br />
  <Row>
    <Col className="demo-col" xs={4} md={6}>A</Col>
    <Col className="demo-col" xs={10} md={6}>B</Col>
    <Col className="demo-col" xs={10} md={12}>C</Col>
  </Row>
</Container>
```

Offset
```js
import Container from '../Container';
import Row from '../Row';

<Container>
  <Row>
    <Col className="demo-col" offset={3} span={3}>A</Col>
    <Col className="demo-col" span={12}>B</Col>
    <Col className="demo-col" span={6}>C</Col>
  </Row>
  <br />
  <Row>
    <Col className="demo-col" span={6}>A</Col>
    <Col className="demo-col" span={6}>B</Col>
    <Col className="demo-col" offset={6} span={6}>C</Col>
  </Row>
</Container>
```

Grow, isAuto, Order, Shrink
```js
import Container from '../Container';
import Row from '../Row';

<Container>
  <Row>
    <Col className="demo-col">A</Col>
    <Col className="demo-col" grow={0}>B</Col>
    <Col className="demo-col">C</Col>
  </Row>
  <br />
  <Row>
    <Col className="demo-col" span={6}>A</Col>
    <Col className="demo-col" isAuto>B</Col>
    <Col className="demo-col" span={6}>C</Col>
  </Row>
  <br />
  <Row>
    <Col className="demo-col" order={3}>A</Col>
    <Col className="demo-col" order={2}>B</Col>
    <Col className="demo-col" order={1}>C</Col>
  </Row>
  <br />
  <Row>
    <Col className="demo-col">A</Col>
    <Col className="demo-col" shrink={0}>B</Col>
    <Col className="demo-col" shrink={1}>C</Col>
  </Row>
</Container>
```

Align Self
```js
import Container from '../Container';
import Row from '../Row';

<Container>
  <Row>
    <Col className="demo-col" align="auto">
      A
    </Col>
    <Col className="demo-col" align="start">
      B
    </Col>
    <Col className="demo-col" align="end">
      C
    </Col>
    <Col className="demo-col" align="center">
      A
    </Col>
    <Col className="demo-col" align="baseline">
      B
    </Col>
    <Col className="demo-col" align="stretch">
      C
    </Col>
    <Col className="demo-col"><div style={{height:'200px'}}></div></Col>
  </Row>
</Container>
```

Mixed on specific breakpoints
```js
import Container from '../Container';
import Row from '../Row';

<Container>
  <Row>
    <Col className="demo-col" sm={{span:6}} md={{span:8,order:3}}>A</Col>
    <Col className="demo-col" sm={{span:12}} md={{span:8,order:1}}>B</Col>
    <Col className="demo-col" sm={{span:6}} md={{span:8,order:2}}>C</Col>
  </Row>
</Container>
```

* <i>The custom class demo-col is defined and used for the Styleguidist only.</i>