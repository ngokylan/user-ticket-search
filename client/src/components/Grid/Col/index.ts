export { default } from './Col';
export * from './type';
