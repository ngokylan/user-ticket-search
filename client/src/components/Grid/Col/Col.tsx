import React from 'react';
import { mapPropsToClass } from './api';
import { ColProps } from './type';
import './Col.scss';

function Col({ children, ...props }: ColProps): any {
  const className = mapPropsToClass('col', props);
  const { role, key } = props;

  return (
    <div className={className} key={key} role={role}>
      {children}
    </div>
  );
}

Col.defaultProps = {
  isAuto: false,
  isFill: false,
  role: 'cell',
};

export default Col;
