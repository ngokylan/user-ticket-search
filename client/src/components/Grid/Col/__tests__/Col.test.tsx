import React from 'react';
import { mount } from 'enzyme';
import Col, { ColProps } from '..';

const defaultProps: ColProps = {
  isAuto: false,
  isFill: false,
  role: 'cell',
};
const mountComponent = (customProps: any = {}) => {
  const props = { ...defaultProps, ...customProps };
  const wrapper = mount(<Col {...props} />);

  return { props, wrapper };
};

describe('Col', () => {
  describe('rendering', () => {
    it('should render component without any crash', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toHaveLength(1);
    });

    it('should be match with snapshot', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toMatchSnapshot();
    });

    it.each([
      ['isAuto', true],
      ['isAuto', false],
      ['span', 2],
      ['order', 2],
      ['offset', 2],
      ['align', 'auto'],
      ['shrink', 2],
      ['grow', 2],
      ['isFill', true],
      ['isFill', false],
      ['xs', 2],
      ['sm', 2],
      ['md', 2],
      ['lg', 2],
      ['xl', 2],
      ['xxl', 2],
      ['xxxl', 2],
      ['key', 2],
      ['className', 'className'],
      ['className', null],
      ['className', undefined],
      ['className', false],
    ])(
      'should be match with snapshot when props %s is %s',
      (propsName, propsValue) => {
        const { wrapper } = mountComponent({ [propsName]: propsValue });
        expect(wrapper).toMatchSnapshot();
      }
    );
  });

  describe('props', () => {
    describe('default', () => {
      it('should have expected default props', () => {
        expect(JSON.stringify(defaultProps)).toBe(
          JSON.stringify(Col.defaultProps)
        );
      });
    });

    describe('children', () => {
      it('should render children', () => {
        const text = 'Hello world';
        const { wrapper } = mountComponent({ children: text });
        expect(wrapper.text()).toBe(text);
      });
    });
  });
});
