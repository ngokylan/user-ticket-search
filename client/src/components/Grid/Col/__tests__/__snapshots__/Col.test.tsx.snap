// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Col rendering should be match with snapshot 1`] = `
<div
  className="col"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props align is auto 1`] = `
<div
  className="col align-self-auto"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props className is className 1`] = `
<div
  className="col className"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props className is false 1`] = `
<div
  className="col"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props className is null 1`] = `
<div
  className="col"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props className is undefined 1`] = `
<div
  className="col"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props grow is 2 1`] = `
<div
  className="col flex-grow-2"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props isAuto is false 1`] = `
<div
  className="col"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props isAuto is true 1`] = `
<div
  className="col col-auto"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props isFill is false 1`] = `
<div
  className="col"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props isFill is true 1`] = `
<div
  className="col flex-fill"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props key is 2 1`] = `
<div
  className="col"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props lg is 2 1`] = `
<div
  className="col col-lg-2"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props md is 2 1`] = `
<div
  className="col col-md-2"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props offset is 2 1`] = `
<div
  className="col offset-2"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props order is 2 1`] = `
<div
  className="col order-2"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props shrink is 2 1`] = `
<div
  className="col flex-shrink-2"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props sm is 2 1`] = `
<div
  className="col col-sm-2"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props span is 2 1`] = `
<div
  className="col col-2"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props xl is 2 1`] = `
<div
  className="col col-xl-2"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props xs is 2 1`] = `
<div
  className="col col-xs-2"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props xxl is 2 1`] = `
<div
  className="col col-xxl-2"
  role="cell"
/>
`;

exports[`Col rendering should be match with snapshot when props xxxl is 2 1`] = `
<div
  className="col col-xxxl-2"
  role="cell"
/>
`;
